Python/thread_nt.h

```diff
@@ -266,6 +266,27 @@ PyThread_exit_thread(void)
     _endthreadex(0);
 }
 
+void _Py_NO_RETURN
+_PyThread_hang_thread(void)
+{
+    MSG msg;
+
+    dprintf(("%lu: PyThread_hang_thread called\n", PyThread_get_thread_ident()));
+    // Still process Windows messages to avoid system deadlock.
+    //
+    // According to MSDN, if the thread has created windows, which can happen
+    // implicitly due to calls to `CoInitialize`, it must process messages,
+    // because message broadcasts, which are sent to all windows in the system,
+    // will deadlock if a thread does not process messages.
+    //
+    // https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep#remarks
+    while (1) {
+        GetMessage(&msg, NULL, 0, 0);
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+    }
```

Comment:
Won't this be a problem if a window procedure is implemented as a Python function (e.g. via PyWin32 or ctypes)?

Also, there's nothing to be done if it's not a GUI process/thread, i.e. when the thread doesn't even have a message queue. In this case it can just call `Sleep(INFINITE)`.

To avoid loading "user32", which converts to a GUI process (something I'd like to avoid/delay in the future, but it's not currently the case), I suggest dynamically checking for a GUI thread as follows:

```C
static BOOL 
is_gui_thread()
{
    HMODULE user32 = GetModuleHandleW(L"user32");
    if (user32 != NULL) {
        FARPROC pf = GetProcAddress(user32, "IsGUIThread");
        if (pf != NULL) {
            typedef BOOL (WINAPI *pfIsGUIThread)(BOOL);
            return ((pfIsGUIThread)pf)(FALSE);
        }
    }
    return FALSE;
}
```
For now, Python implicitly links with "user32", so the dynamic use of `GetModuleHandleW()` and `GetProcAddress()` doesn't matter. However, the `IsGUIThread(FALSE)` call does matter. A thread doesn't get converted to a GUI thread until it uses any GUI related functions, such as creating a window. Currently the main thread is converted at startup during the static import of "user32". But initially new threads are not GUI threads.

Reply:
I made the change as you suggest.

You have a good point regarding a window procedure implemented in Python --- I guess it would attempt to re-acquire the GIL, and we would have unbounded recursion back into the hang function, until we overflow the stack.

Ideally PyWin32 would be changed to use `PyThread_TryAcquireFinalizeBlock` or `PyGILState_TryAcquireFinalizeBlockAndGil` rather than `PyGILState_Ensure`.  But still we don't want to cause problems for existing code.  I suppose one option could be to detect re-entering of `_PyThread_hang_thread` and just Sleep in that case.

Reply:
I updated it to guard against re-entrant calls.